### PR TITLE
Compressor now uses GZip encoding

### PR DESCRIPTION
**Pull request by adamretter**
Previously Zip file encoding was used, which has quite some overhead - due to having a directory layout format inside the archive. Now the just the data stream is compressed with GZip, which has a much smaller overhead as it is not an archive format.

NOTE: This introduces a non-backwards compatible change to the XML-RPC API (and by extension the XML:DB API and Java Admin Client).

### CODE REVIEW

**Comment by duncdrum** [comment]
@adam i feel we should we add a note about this to [the docs](http://www.exist-db.org/exist/apps/doc/devguide_xmlrpc.xml?field=all&id=D3.3.2), despite there not being one right now

**Comment by adamretter** [comment]
@duncdrum it at least needs to go in the release notes as a non-backwards compatible change

**Comment by adamretter** [comment]
@duncdrum It is not configurable. It used to be zip, now it is gzip.

**Comment by duncdrum** [comment]
@adamretter I understand that; but users can call `upload `vs `uploadCompressed` and the second doesn't appear in the docs about what users can or can't call.

File: src/org/exist/util/Compressor.java [hunk 1]
```diff
@@ -1,94 +1,112 @@
...
-        try (final FastByteArrayOutputStream baos = new FastByteArrayOutputStream(length);
-                final ZipOutputStream gzos = new ZipOutputStream(baos)) {
-            gzos.setMethod(ZipOutputStream.DEFLATED);
-            gzos.putNextEntry(new ZipEntry(length + ""));
-            gzos.write(whatToCompress, 0, length);
-            gzos.closeEntry();
+    public static byte[] compress(final byte[] buf, final int len) throws IOException {
+        try (final FastByteArrayOutputStream baos = new FastByteArrayOutputStream(len);
+                final GZIPOutputStream gzos = new GZIPOutputStream(baos)) {
+            gzos.write(buf, 0, len);
```
Comment (dizzzz) [inline src/org/exist/util/Compressor.java]: `gzos.flush()` is required here, the `close()` will not do that, strangely.

File: src/org/exist/util/Compressor.java [hunk 2]
```diff
@@ -1,94 +1,112 @@
...
+     * @exception IOException if an error occurs
+     */
+    public static void uncompress(final byte[] buf, final OutputStream os) throws IOException {
+        try (final FastByteArrayInputStream bais = new FastByteArrayInputStream(buf);
+             final InputStream gzis = new GZIPInputStream(bais)) {
+            final byte[] tmp = new byte[4096];
+            int read;
+            while ((read = gzis.read(tmp)) != -1) {
+                os.write(tmp, 0, read);
+            }
```
Comment (dizzzz) [inline src/org/exist/util/Compressor.java]: flush buffer here?

File: src/org/exist/util/Compressor.java (hunk 1 above)
Comment (adamretter) [reply src/org/exist/util/Compressor.java]: Are you certain? Looking at the code, it appears to me that `finish()` (below) will do a very similar thing to a `flush()`.

File: src/org/exist/util/Compressor.java (hunk 2 above)
Comment (adamretter) [reply src/org/exist/util/Compressor.java]: I don't think it is needed.

File: src/org/exist/util/Compressor.java (hunk 1 above)
Comment (dizzzz) [reply src/org/exist/util/Compressor.java]: ah crap, sorry, i meant `finish()`. I guess it is the heat.

File: src/org/exist/util/Compressor.java (hunk 2 above)
Comment (dizzzz) [reply src/org/exist/util/Compressor.java]: Hmmmmmmm ok

**Pull request merged**